helpers/metrics.py

```diff
@@ -0,0 +1,172 @@
+    fn_gdf_temp.loc[:, 'geometry'] = fn_gdf_temp.label_geom
+
+    actual_matches_gdf['IOU'] = actual_matches_gdf.IOU.round(3)
+
+    # Test that it has the right class (id starting at 1 for labels and at 0 for detections)
+    condition = actual_matches_gdf.label_class == actual_matches_gdf.det_class+1
+    tp_gdf = actual_matches_gdf[condition].reset_index(drop=True)
+    mismatched_classes_gdf = actual_matches_gdf[~condition].reset_index(drop=True)
+    mismatched_classes_gdf.drop(columns=['x', 'y', 'z', 'dataset_right', 'label_geom'], errors='ignore', inplace=True)
+    mismatched_classes_gdf.rename(columns={'dataset_left': 'dataset'}, inplace=True)
+
+
+    # FALSE POSITIVES
+    fp_gdf = left_join[left_join.label_id.isna()].copy()
+    assert(len(fp_gdf[fp_gdf.duplicated()]) == 0)
+    fp_gdf = pd.concat([fp_gdf_temp, fp_gdf], ignore_index=True)
+    fp_gdf.drop(columns=_labels_gdf.columns.to_list() + ['index_right', 'dataset_right', 'label_geom', 'IOU'], errors='ignore', inplace=True)
+    fp_gdf.rename(columns={'dataset_left': 'dataset'}, inplace=True)
+    
+    # FALSE NEGATIVES
+    right_join = gpd.sjoin(_dets_gdf, _labels_gdf, how='right', predicate='intersects', lsuffix='left', rsuffix='right')
+    fn_gdf = right_join[right_join.score.isna()].copy()
+    fn_gdf.drop_duplicates(subset=['label_id', 'tile_id'], inplace=True)
+    fn_gdf = pd.concat([fn_gdf_temp, fn_gdf], ignore_index=True)
+    fn_gdf.drop(columns=_dets_gdf.columns.to_list() + ['dataset_left', 'index_right', 'x', 'y', 'z', 'label_geom', 'IOU', 'index_left'], errors='ignore', inplace=True)
+    fn_gdf.rename(columns={'dataset_right': 'dataset'}, inplace=True)
+    
+    return tp_gdf, fp_gdf, fn_gdf, mismatched_classes_gdf
+
+
+def get_metrics(tp_gdf, fp_gdf, fn_gdf, mismatch_gdf, id_classes=0):
+    """Determine the metrics based on the TP, FP and FN
+
+    Args:
+        tp_gdf (geodataframe): true positive detections
+        fp_gdf (geodataframe): false positive detections
+        fn_gdf (geodataframe): false negative labels
+        mismatch_gdf (geodataframe): labels and detections intersecting with a mismatched class id
+        id_classes (list): list of the possible class ids. Defaults to 0.
+    
+    Returns:
+        tuple: 
+            - dict: precision for each class
+            - dict: recall for each class
+            - float: precision;
+            - float: recall;
+            - float: f1 score.
+    """
+    
+    by_class_dict = {key: None for key in id_classes}
+    tp_k = by_class_dict.copy()
+    fp_k = by_class_dict.copy()
+    fn_k = by_class_dict.copy()
+    p_k= by_class_dict.copy()
+    r_k= by_class_dict.copy()
+    
+    for id_cl in id_classes:
+
+        tp_count= len(tp_gdf[tp_gdf.det_class==id_cl])
+        fp_count= len(fp_gdf[fp_gdf.det_class==id_cl]) + len(mismatch_gdf[mismatch_gdf.det_class == id_cl])
```

Comment:
Are we sure there are no duplicate lines between fp_gdf and mismatch_gdf?

Reply:
Yes, since fp_gdf are detections where there are no matching label and msimatch_gdf have a matching label, only with a different category.